### ISSUE TITLE
Add configuration option to allow running calico-node as a NetworkPolicy controller only

### ISSUE THREAD

**Issue by juanfresia** [projectcalico/felix#2659]
Hi everyone! This is not a bug, but more like a feature request of sorts.

## tl;dr

We want to propose adding a configuration to disable any interaction between `calico-node` (through `felix`) and the routing tables. This is to allow running Calico as a NetworkPolicy controller only.

## Longer version

We want to know how hard would it be to modify `felix` to allow running `calico-node` in a true Network Policy only mode. That is, to have `calico-node` only interact with `iptables` and `ipset` **without ever modifying any routing table**.

We managed to have it running that way by replacing [this assignment](https://github.com/projectcalico/felix/blob/092e3b76f40c0eba36c94a78752abc2856878509/dataplane/linux/int_dataplane.go#L676) with a dummy interface (only here, because we needed it for ipv4 and not ipv6). That prevented `calico-node` to touching the routing table, while letting it configure `iptables` normally.

As a side note, the names of the interfaces still have to match what Calico expects (both the prefix, and the hashed Pod name/namespace)

Of course, that was only a quick PoC; but a proper change could be considered by adding a configuration parameter to disable the route table sync (maybe `EnableRouteTableManager=false`). The purpose of this issue is to discuss:

0) Is there already a better way of running `calico-node` as Network Policy only mode that we've may have overlooked?
1) Do you think it would be worthwhile to have such configuration to disable all interaction with the routing table?
2) Do you see any issues with this approach?
3) Would it suffice to track down usages of `routeTableSyncer` and `routeTable` interfaces and replace them with noop ones? There are not many of these, and disabling should be straightforward.

We believe that this could open the door to more users that only want/need a NetworkPolicy manager; since there are not many out of the box solutions for *only that*.

If you think it's viable, I'd love to work on a PR and send it your way!

Let us know any thoughts you may have on this, all feedback is appreciated.
I'll leave more context below, fwiw.

## More context

We are running a k8s cluster with a homemade CNI plugin (the details of that are not too relevant, but we have some specific requirement that require us to have multiple IPs per Pod). This plugin runs in a similar fashion to the `ptp` standard plugin: it creates routes in the host's routing table and expects someone else to propagate the routes across the cluster. We are using `frr` for that.

Up to recently we were using `kube-router` as an implementation for NetworkPolicies, but it turned out to not scale well with the total number of Policies. So we looked for alternatives and one of them was Calico.

We wanted to run **Calico as a NetworkPolicy controller only**, and found out that there was no easy way of doing so.

We tried [calico-policy-only.yaml](https://docs.projectcalico.org/v3.17/manifests/calico-policy-only.yaml), and deleted `calico-kube-controllers` from there. Up to that point we had one `calico-node` DaemonSet, and one `calico-typha` Deployment; and with that NetworkPolicies were implemented by Calico. (The folks at cybozu-go had [a similar situation](https://github.com/cybozu-go/neco-apps/blob/main/docs/maintenance.md#network-policy-calico), we use it as a base)

However, that setup also messed up with our routes, particularly it deleted any route we had for those "extra IPs" (i.e. any IP we assign to the Pod via CNI, that does not reflect through the k8s apiserver). In particular, we noticed that the `endpoint manager` was active and attempting a sync of the routing table with every pod created/destroyed. This is of course desirable if Calico is used for both routing and NetworkPolicies, but we already had our solution for routing that we sadly cannot change. So we tried looking for a way to disable that behaviour.

This is an example of the logs of `calico-node` deleting routes. Something to notice here is that we changed the interface names of all our veth to match what Calico expects. This lead us to having Calico **only delete the extra IPs routes**. Any route that matched with the Pod IP will be validated by Calico and left intact.

```
2021-01-25 20:24:33.019 [INFO][64] felix/route_table.go 865: Remove old route dest=10.228.200.32/32 ifaceName="veth74985cde457" ifaceRegex="^veth.*" ipVersion=0x4 routeProblems=[]string{"unexpected route"}
2021-01-25 20:24:33.019 [INFO][64] felix/conntrack.go 90: Removing conntrack flows ip=10.228.200.32
```

We didn't find a way of disabling this behavior from the configuration; so we took a look at the code to see how hard would it be to run `calico-node` but with the endpoint manager disabled. It looks like that the code that the manager that controls the routing table is initialized [here](https://github.com/projectcalico/felix/blob/092e3b76f40c0eba36c94a78752abc2856878509/dataplane/linux/int_dataplane.go#L676). We attempted replacing the current implementation `routetable` with a dummy one that did nothing and that seemed to solve our issues.

So the origin of this issue was to see if it would be worthwhile to work this into a proper configuration, if anyone else would be interested in that.

**Comment by caseydavenport** [comment]
@fasaxc what do you think? Seems more or less reasonable to me.

> However, that setup also messed up with our routes, particularly it deleted any route we had for those "extra IPs" (i.e. any IP we assign to the Pod via CNI, that does not reflect through the k8s apiserver)

It's too bad there is no way to pass multiple IPs back to k8s. It has a "status.PodIPs" field but it's really limited in scope to just dual-stack, so doesn't support more than 1 of each address family. If it did, this would be a non issue, since Felix would be aware of the addresses on the pod and thus wouldn't try to reconcile the routes away.

**Comment by fasaxc** [comment]
I think you want the `RemoveExternalRoutes` config parameter.  If you set that to `false`, we'll stop removing your routes and only maintain our own.  Probably enough for your use case.

I don't think we can do policy-only in the general case; the CNI plugin needs to use our veth naming scheme or we won't be able to find the veths and secure them.  There are also other subtle requirements we have on the CNI plugin.  For example, it must be routed and not use a bridge.  Ideally, it should set the Calico podIPs annotation on the pod to avoid slow pod networking start up etc.

I suppose we could learn the veth name from the routes but then we have to parse arbitrary routes, which is painful (and if the routes go away, we can't fix them).

**Comment by juanfresia** [comment]
> It's too bad there is no way to pass multiple IPs back to k8s

@caseydavenport  Yes, that's too bad; we hit that fact too. It kinda limits what CNI plugins can do. 😕 

@fasaxc  As I tried to explain in the issue, using the `InterfacePrefix` we can make Calico find the interfaces and create Network Policies around them (since it uses the interface name instead of the destination IP to match traffic). However, if the complete interface name does not match the expected name (hash of Pod Name and Namespace), it _won't recongnize_ any of the routes, and delete _all of them_. That's the behaviour we saw.

> I think you want the RemoveExternalRoutes config parameter

So we ended up changing the naming of our interfaces to match what Calico expects, and that worked _except_ for the "extra IPs". I didn't mentioned, but we found and have set `RemoveExternalRoutes` to `false`, however that changed nothing and the behaviour was the same. I don't think any route matching a `veth` whose prefix Calico recognizes counts as external, but please correct me if I'm wrong.

> I don't think we can do policy-only in the general case.

I definitely agree that the general case would be way harder; the conditions I'd suggest for the policy-only to work with other CNIs could be:
1) The CNI names the interfaces consistently with what Calico expects
2) The CNI uses a one-interface-per-pod, schema (like you said, `veth` and no `bridge`s)

With those conditions it would be much more sensible to implement, but maybe they are too restrictive.

> Seems more or less reasonable to me.

Again, if this makes sense (conditions included), I'll galdly work on the needed changes.

**Comment by fasaxc** [comment]
`RemoveExternalRoutes` spots external routes based on the protocol of the route.  If you set a different protocol for your routes then we'll ignore them. Or, you can set felix to use a different protocol with `DeviceRouteProtocol`
